### PR TITLE
fix json path for external dataset/subscription ids on actions & trig…

### PR DESCRIPTION
**Pull request by maybeiambatman**
…gers #2

### CODE REVIEW

**Comment by zroadhouse-rmn** [comment]
Can you add an integration test that utilizes these features? Something in tests/test_sync_manager.py would be useful.

**Comment by zroadhouse-rmn** [comment]
Also, can you bump the dev # in the version? (i.e. to 1.1.dev3)

File: dartclient/core.py [hunk 1]
```diff
@@ -276,6 +291,20 @@ def find_dataset(self, dataset_name):
...
+        Find the dataset by name
+
+        :param subcription_name: the subscription name
+        :param dataset: the connected dataset
+        :return: the subscription object or None if not found
+        """
+        response = self.client.Subscription.listSubscriptions(
+            filters=self.filter_by(name=subscription_name)).result()
+        if response.total > 1:
+            raise Exception("More than one subcription object found.")
```
Comment (zroadhouse-rmn) [inline dartclient/core.py]: "subcription" is misspelled.

File: dartclient/core.py [hunk 2]
```diff
@@ -276,6 +291,20 @@ def find_dataset(self, dataset_name):
             raise Exception("More than one dataset object found.")
         return response.results[0] if response.total > 0 else None
 
+    def find_subscription(self, subscription_name):
+        """
+        Find the dataset by name
+
+        :param subcription_name: the subscription name
+        :param dataset: the connected dataset
```
Comment (zroadhouse-rmn) [inline dartclient/core.py]: There's no dataset parameter - pls remove.

File: dartclient/core.py [hunk 3]
```diff
@@ -276,6 +291,20 @@ def find_dataset(self, dataset_name):
             raise Exception("More than one dataset object found.")
         return response.results[0] if response.total > 0 else None
 
+    def find_subscription(self, subscription_name):
+        """
+        Find the dataset by name
```
Comment (zroadhouse-rmn) [inline dartclient/core.py]: Find the subscription by name - likely copy/paste err

File: dartclient/core.py [hunk 4]
```diff
@@ -408,9 +447,13 @@ def sync_action(self, action_name, workflow, callback, dataset=None, subscriptio
         if action:
             action = callback(action)
             if dataset:
-                action.args.dataset_id = dataset.id
+                action.data.args = {
```
Comment (zroadhouse-rmn) [inline dartclient/core.py]: You are going to override the existing data.args dictionary as set by callback if you assign a dictionary here to action.data.args. This should probably be something like:
```
if not action.data.args:
    action.data.args = {}
action.data.args['dataset_id'] = dataset.id
```

File: dartclient/core.py [hunk 5]
```diff
@@ -408,9 +447,13 @@ def sync_action(self, action_name, workflow, callback, dataset=None, subscriptio
...
+                    'dataset_id': dataset.id
+                }
             if subscription:
-                action.args.subscription_id = subscription.id
+                action.data.args = {
```
Comment (zroadhouse-rmn) [inline dartclient/core.py]: Same as previous comment.

File: dartclient/core.py [hunk 6]
```diff
@@ -421,9 +464,13 @@ def sync_action(self, action_name, workflow, callback, dataset=None, subscriptio
                 action.data.state = action_state
             action = callback(action)
             if dataset:
-                action.args.dataset_id = dataset.id
+                action.data.args = {
```
Comment (zroadhouse-rmn) [inline dartclient/core.py]: Same as previous comment.

File: dartclient/core.py [hunk 7]
```diff
@@ -421,9 +464,13 @@ def sync_action(self, action_name, workflow, callback, dataset=None, subscriptio
...
+                    'dataset_id': dataset.id
+                }
             if subscription:
-                action.args.subscription_id = subscription.id
+                action.data.args = {
```
Comment (zroadhouse-rmn) [inline dartclient/core.py]: Same as previous comment.

File: dartclient/core.py [hunk 8]
```diff
@@ -440,7 +487,9 @@ def sync_trigger(self, trigger_name, workflow, callback, subscription=None):
         if trigger:
             trigger = callback(trigger)
             if subscription:
-                trigger.args.subcription_id = subscription.id
+                trigger.data.args = {
```
Comment (zroadhouse-rmn) [inline dartclient/core.py]: Same as previous comment.

File: dartclient/core.py [hunk 9]
```diff
@@ -451,14 +500,16 @@ def sync_trigger(self, trigger_name, workflow, callback, subscription=None):
             if workflow:
                 trigger.data.workflow_ids = [workflow.id]
             if subscription:
-                trigger.args.subcription_id = subscription.id
+                trigger.data.args = {
```
Comment (zroadhouse-rmn) [inline dartclient/core.py]: Same as previous comment.

File: dartclient/core.py [hunk 10]
```diff
@@ -477,3 +528,22 @@ def sync_dataset(self, dataset_name, callback):
             response = self.client.Dataset.createDataset(
                 dataset=dataset).result()
             return response.results
+
+    def sync_subscription(self, subscription_name, dataset, callback):
+        """
+        Synchronize a subscription with Dart.
+
+        :param subscription_name: The name of the subscription
+        :param callback: A function with a signature (subscription) => subscription
```
Comment (zroadhouse-rmn) [inline dartclient/core.py]: Please add a param for dataset

File: dartclient/core.py [hunk 11]
```diff
@@ -477,3 +528,22 @@ def sync_dataset(self, dataset_name, callback):
...
+        :return: The created or updated subscription
+        """
+        subscription = self.find_subscription(subscription_name)
+        if subscription:
```
Comment (zroadhouse-rmn) [inline dartclient/core.py]: Interesting - do you have to clean a subscription? There's no way to update it?

File: tests/test_sync_manager.py [hunk 12]
```diff
@@ -98,22 +106,39 @@ def __init__(self, client):
     def clean(self):
         self.sync_manager.clean_datastore(self.sync_manager.find_datastore(
             self.DATASTORE1_NAME, self.DATASTORE1_STATE))
+        self.sync_manager.clean_datastore(self.sync_manager.find_datastore(
+            self.DATASTORE2_NAME, self.DATASTORE2_STATE))
+        self.sync_manager.clean_subscription(
```
Comment (zroadhouse-rmn) [inline tests/test_sync_manager.py]: Are subscriptions owed by the target dataset? If so, if you clean that dataset, does it (or should it) clean the associated subscriptions?

File: tests/test_sync_manager.py [hunk 13]
```diff
@@ -172,26 +197,71 @@ def define_dataset1(self, dataset):
...
+        datastore.data.concurrency = 1
+        datastore.data.engine_name = 'no_op_engine'
+        return datastore
+
+    def define_workflow2(self, workflow):
+        workflow.data.concurrency = 1
+        workflow.data.engine_name = 'no_op_engine'
+        return workflow
+
+    def define_action3(self, action):
```
Comment (zroadhouse-rmn) [inline tests/test_sync_manager.py]: Can you define an action with args to test the earlier case that I mentioned about the args dictionary getting overwritten?

File: dartclient/core.py (hunk 4 above)
Comment (maybeiambatman) [inline dartclient/core.py]: Good catch!

File: tests/test_sync_manager.py (hunk 12 above)
Comment (maybeiambatman) [inline tests/test_sync_manager.py]: Yup. That's a good idea. I will make it so that clean dataset also cleans up the subscriptions associated with it.

**Comment by maybeiambatman** [comment]
Integrations checks have passed. There is still a weird lingering issue around creating Triggers. When Triggers are created they attach an extra tag which appears to be a random string. Since tags won't break anything, I will merge this in.

**Pull request merged**